sound/soc/sof/debug.c

```diff
@@ -336,6 +336,103 @@ static int sof_debug_ipc_flood_test(struct snd_sof_dev *sdev,
 }
 #endif
 
+#if IS_ENABLED(CONFIG_SND_SOC_SOF_DEBUG_IPC_MSG_INJECTOR)
+static ssize_t msg_inject_read(struct file *file, char __user *buffer,
+			       size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct sof_ipc_reply *rhdr = dfse->msg_inject_rx;
+
+	if (!rhdr->hdr.size || !count || *ppos)
+		return 0;
+
+	if (count > rhdr->hdr.size)
+		count = rhdr->hdr.size;
+
+	if (copy_to_user(buffer, dfse->msg_inject_rx, count))
+		return -EFAULT;
+
+	*ppos += count;
+	return count;
+}
+
+static ssize_t msg_inject_write(struct file *file, const char __user *buffer,
+				size_t count, loff_t *ppos)
+{
+	struct snd_sof_dfsentry *dfse = file->private_data;
+	struct snd_sof_dev *sdev = dfse->sdev;
+	struct sof_ipc_cmd_hdr *hdr = dfse->msg_inject_tx;
+	size_t size;
+	int ret, err;
+
+	if (!count || *ppos || count > SOF_IPC_MSG_MAX_SIZE)
+		return 0;
+
+	size = simple_write_to_buffer(dfse->msg_inject_tx, count, ppos, buffer,
+				      count);
+	memset(dfse->msg_inject_tx + count, 0, SOF_IPC_MSG_MAX_SIZE - count);
+
+	ret = pm_runtime_get_sync(sdev->dev);
+	if (ret < 0 && ret != -EACCES) {
+		dev_err_ratelimited(sdev->dev, "%s: DSP resume failed: %d\n",
+				    __func__, ret);
+		pm_runtime_put_noidle(sdev->dev);
+		goto out;
+	}
+
+	/* send the message */
+	memset(dfse->msg_inject_rx, 0, SOF_IPC_MSG_MAX_SIZE);
+	ret = sof_ipc_tx_message(sdev->ipc, hdr->cmd, dfse->msg_inject_tx, count,
+				 dfse->msg_inject_rx, SOF_IPC_MSG_MAX_SIZE);
+
+	pm_runtime_mark_last_busy(sdev->dev);
+	err = pm_runtime_put_autosuspend(sdev->dev);
+	if (err < 0)
+		dev_err_ratelimited(sdev->dev, "%s: DSP idle failed: %d\n",
+				    __func__, err);
+
+	/* return size if test is successful */
+	if (ret >= 0)
+		ret = size;
+
+out:
+	return ret;
+}
+
+static const struct file_operations msg_inject_fops = {
+	.open = simple_open,
+	.read = msg_inject_read,
+	.write = msg_inject_write,
+	.llseek = default_llseek,
+};
+
+static int snd_sof_debugfs_msg_inject_item(struct snd_sof_dev *sdev,
+					   const char *name, mode_t mode,
+					   const struct file_operations *fops)
+{
+	struct snd_sof_dfsentry *dfse;
+
+	dfse = devm_kzalloc(sdev->dev, sizeof(*dfse), GFP_KERNEL);
+	if (!dfse)
+		return -ENOMEM;
+
+	/* pre allocate the tx and rx buffers */
+	dfse->msg_inject_tx = devm_kzalloc(sdev->dev, SOF_IPC_MSG_MAX_SIZE, GFP_KERNEL);
```

Comment:
SOF_IPC_MSG_MAX_SIZE is very limited and only works for small size of ipc message. you can use sdev->host_box.size which is the max size of ipc message.  For IPC4 case, we will not use SOF_IPC_MSG_MAX_SIZE for above reason

Reply:
with such change, you need to refine your inject_write function

Reply:
@RanderWang, the IPC core can not handle anything larger than `SOF_IPC_MSG_MAX_SIZE`, because:
https://github.com/thesofproject/linux/blob/5a6839b58e44edb5a0f5811d3573bca26cbf1217/sound/soc/sof/ipc.c#L931
and
https://github.com/thesofproject/linux/blob/5a6839b58e44edb5a0f5811d3573bca26cbf1217/sound/soc/sof/ipc.c#L936

This is a limitation we have at the moment.

Reply:
Not able to follow your comment @RanderWang, can you elaborate?

Reply:
For current sof, we split big ipc msg into small batch of ipc msg, please check sof_set_get_large_ctrl_data, which is for filter data or other byte array.  And for ipc4, the max ipc size will be not SOF_IPC_MSG_MAX_SIZE. I will create a PR to address ipc3 & ipc4.

Reply:
@ujfalusi  you can ignore ipc4 since I can make change in my PR based on your work, but current large ipc message support should be included

Reply:
@RanderWang, the API for `sof_set_get_large_ctrl_data()` is far from a generic interface to be able to pass any sized IPC message and afaik it can only be done for `SOF_IPC_GLB_COMP_MSG` type of messages.
I guess we could try to interpret the injected message which is larger than SOF_IPC_MSG_MAX_SIZE as `struct sof_ipc_ctrl_data` and try to break it down.
This would require validation of the data given to be injected to be able to craft split messages. To complicate this, I have received a binary message which supposed to cause issues on the firmware side, the file is 388 bytes long, so a long message but the data within does not make sense, does not map to `struct sof_ipc_ctrl_data` so we would not be able to pass it anyways.

It might be possible to get it working with valid binary if I can get hold of one...

Reply:
yes, I know this is a corner case and make it much more complex. I also don't have any experience for such big ipc message. @ranj063 , do you have any idea ?

Reply:
One of the example binary message file I have received starts with:
```
00000000  11 00 00 00 2e 01 02 50  38 74 86 13 a3 73 c4 18  |.......P8t...s..|
00000010  93 ee 54 22 03 e7 d7 70  aa db 98 43 a8 30 e8 da  |..T"...p...C.0..|
00000020  52 46 81 24 51 34 32 39  34 39 34 39 36 37 35 35  |RF.$Q42949496755|
00000030  34 c8 07 f1 5f dd 82 da  c9 2d 82 e9 be bc f1 ed  |4..._....-......|
```
The file is 388 bytes long, so it supposed to be a long message. Even the first u32 which must be the size in the header looks suspicious with it's `17`, but the rest is not saner either if I map it to `struct sof_ipc_ctrl_data cdata`:
```
cdata {
	.rhdr {
		.hdr {
			.size = 17;
			.cmd = 0x5002012e;
		};
		.error = 0x13867438;
	};
	.comp_id = 0x18c473a3;
	.type = 0x2254ee93;
	.cmd = 0x70d7e703;
	.index = 0x4398dbaa;
	.buffer {
		.hdr {
			.size = 0xdae830a8;
			.cmd = 0x24814652;
		};
		.phy_addr = 0x39323451
		.pages = 0x39343934;
		.size = 0x35353736;
		.reserved[3];
	};
	.num_elems = 0xedf1bcbe;
	.elems_remaining = 0xd60c29be;
	.msg_index = 0x7ed5ffd4;
....
};
```
It juts makes no sense and can not be parsed, it is supposed to be a `SOF_IPC_GLB_COMP_MSG | SOF_IPC_COMP_GET_VALUE`, but I don't know what it is and how to interpret and handle it in a Linux safe way...

Reply:
@RanderWang @ujfalusi I think we should keep this simple like it is now and skip large IPCs. They are specifically for comp data and we already have the cset tlv mechanism if we want to validate large IPCs already.